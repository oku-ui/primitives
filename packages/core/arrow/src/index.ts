export {
  OkuArrow,
  arrowProps,
} from './arrow'

export type {
  ArrowProps,
  ArrowElement,
  ArrowNaviteElement,
} from './arrow'

export const test = 1
