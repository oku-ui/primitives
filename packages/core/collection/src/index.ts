export {
  createCollection,
  collectionProps,
} from './collection'

export type { CollectionElement, CollectionPropsType } from './collection'
