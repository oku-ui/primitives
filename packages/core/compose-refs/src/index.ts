export { useComposedRefs, composeRefs } from './compose-refs'
