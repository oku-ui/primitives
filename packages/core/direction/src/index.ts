export {
  useDirection,
  OkuDirectionProvider,
} from './Direction'

export type {
  Direction,
} from './Direction'
