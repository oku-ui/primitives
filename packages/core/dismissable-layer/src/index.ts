export * from './DismissableLayerBranch'

export {
  OkuDismissableLayer,
} from './DismissableLayer'

export type {
  DismissableLayerProps,
  DismissableLayerEmits,
} from './props'

export {
  dismissableLayerProps,
} from './props'
