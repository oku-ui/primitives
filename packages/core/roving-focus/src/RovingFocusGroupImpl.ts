import { computed, defineComponent, h, mergeProps, onBeforeUnmount, onMounted, reactive, ref, toRefs } from 'vue'
import { reactiveOmit, useComposedRefs, useControllable, useForwardRef } from '@oku-ui/use-composable'

import { Primitive, primitiveProps } from '@oku-ui/primitive'
import { composeEventHandlers } from '@oku-ui/utils'
import { focusFirst } from './utils'
import type { RovingFocusGroupImplNaviteElement } from './props'
import { ENTRY_FOCUS, EVENT_OPTIONS, rovingFocusGroupImplProps, rovingFocusProvider, scopedProps, useCollection } from './props'

const RovingFocusGroupImpl = defineComponent({
  name: 'OkuRovingFocusGroupImpl',
  inheritAttrs: false,
  props: {
    ...rovingFocusGroupImplProps.props,
    ...primitiveProps,
    ...scopedProps,
  },
  emits: rovingFocusGroupImplProps.emits,
  setup(props, { attrs, slots, emit }) {
    const {
      scopeOkuRovingFocusGroup,
      orientation,
      loop,
      dir,
      currentTabStopId: currentTabStopIdProp,
      defaultCurrentTabStopId,
      ...groupProps
    } = toRefs(props)
    const _reactive = reactive(groupProps)
    const reactiveProupProps = reactiveOmit(_reactive, (key, _value) => key === undefined)

    const buttonRef = ref<HTMLDivElement | null>(null)
    const forwardedRef = useForwardRef()
    const composedRefs = useComposedRefs(buttonRef, forwardedRef)

    const { state: currentTabStopId, updateValue: updateCurrentTabStopId } = useControllable({
      prop: computed(() => currentTabStopIdProp.value),
      defaultProp: computed(() => defaultCurrentTabStopId.value),
      onChange: (result: any) => {
        emit('currentTabStopIdChange', result)
      },
      initialValue: null,
    })

    const isTabbingBackOut = ref(false)
    const getItems = useCollection(scopeOkuRovingFocusGroup.value)
    const isClickFocusRef = ref(false)
    const focusableItemsCount = ref(0)

    onMounted(() => {
      const node = buttonRef.value
      if (node) {
        node.addEventListener(ENTRY_FOCUS, (event) => {
          emit('entryFocus', event)
        })
      }
    })

    onBeforeUnmount(() => {
      const node = buttonRef.value
      if (node) {
        node.removeEventListener(ENTRY_FOCUS, (event) => {
          emit('entryFocus', event)
        })
      }
    })

    rovingFocusProvider({
      scope: scopeOkuRovingFocusGroup.value,
      orientation,
      dir,
      loop,
      currentTabStopId: computed(() => currentTabStopId.value || null),
      onItemFocus: (tabStopId: string) => {
        updateCurrentTabStopId(tabStopId)
      },
      onItemShiftTab: () => {
        isTabbingBackOut.value = true
      },
      onFocusableItemAdd: () => {
        focusableItemsCount.value++
      },
      onFocusableItemRemove: () => {
        focusableItemsCount.value--
      },
    })

    return () => h(Primitive.div, {
      'tabindex': isTabbingBackOut.value || focusableItemsCount.value === 0 ? -1 : 0,
      'data-orientation': orientation?.value,
      ...mergeProps(attrs, reactiveProupProps),
      'ref': composedRefs,
      'style': {
        outline: 'none',
        ...attrs.style as any,
      },
      'onMousedown': composeEventHandlers<MouseEvent>((e) => {
        emit('mousedown', e)
      }, () => {
        isClickFocusRef.value = true
      }),
      'onFocus': composeEventHandlers<FocusEvent>((e) => {
        emit('focus', e)
      }, (event: FocusEvent) => {
        // We normally wouldn't need this check, because we already check
        // that the focus is on the current target and not bubbling to it.
        // We do this because Safari doesn't focus buttons when clicked, and
        // instead, the wrapper will get focused and not through a bubbling event.
        const isKeyboardFocus = !isClickFocusRef.value

        if (event.target === event.currentTarget && isKeyboardFocus && !isTabbingBackOut.value) {
          const entryFocusEvent = new CustomEvent(ENTRY_FOCUS, EVENT_OPTIONS)
          event.currentTarget?.dispatchEvent(entryFocusEvent)
          if (!entryFocusEvent.defaultPrevented) {
            const items = getItems().filter(item => item.focusable)
            const activeItem = items.find(item => item.active)
            const currentItem = items.find(item => item.id === currentTabStopId.value)
            const candidateItems = [activeItem, currentItem, ...items].filter(
              Boolean,
            ) as typeof items
            const candidateNodes = candidateItems.map(item => item.ref.value)
            focusFirst(candidateNodes)
          }
        }

        isClickFocusRef.value = false
      }),
      'onBlur': composeEventHandlers<FocusEvent>((e) => {
        emit('blur', e)
      }, () => {
        isTabbingBackOut.value = false
      }),
    }, slots.default?.())
  },
})

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuRovingFocusGroupImpl = RovingFocusGroupImpl as typeof RovingFocusGroupImpl &
(new () => {
  $props: RovingFocusGroupImplNaviteElement
})
