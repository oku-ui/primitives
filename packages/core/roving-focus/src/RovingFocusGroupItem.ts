import { computed, defineComponent, h, mergeProps, onBeforeUnmount, onMounted, reactive, toRefs } from 'vue'
import { reactiveOmit, useForwardRef, useId } from '@oku-ui/use-composable'

import { Primitive } from '@oku-ui/primitive'

import { composeEventHandlers } from '@oku-ui/utils'
import type { RovingFocusGroupItemNaviteElement } from './props'
import { CollectionItemSlot, ITEM_NAME, rovingFocusItemProps, scopedProps, useCollection, useRovingFocusInject } from './props'
import { focusFirst, getFocusIntent, wrapArray } from './utils'

const rovingFocusGroupItem = defineComponent({
  name: ITEM_NAME,
  components: {
    CollectionItemSlot,
  },
  inheritAttrs: false,
  props: {
    ...rovingFocusItemProps.props,
    ...scopedProps,
  },
  emits: rovingFocusItemProps.emits,
  setup(props, { attrs, slots, emit }) {
    const {
      scopeOkuRovingFocusGroup,
      focusable,
      active,
      tabStopId,
      ...itemProps
    } = toRefs(props)
    const _reactive = reactive(itemProps)
    const reactiveItemProps = reactiveOmit(_reactive, (key, _value) => key === undefined)

    const autoId = useId()
    const id = computed(() => tabStopId.value || autoId)
    const inject = useRovingFocusInject(ITEM_NAME, scopeOkuRovingFocusGroup.value)
    const isCurrentTabStop = computed(() => inject.currentTabStopId.value === id.value)
    const getItems = useCollection(scopeOkuRovingFocusGroup.value)
    const forwardedRef = useForwardRef()

    onMounted(() => {
      if (props.focusable)
        inject.onFocusableItemAdd()
    })

    onBeforeUnmount(() => {
      if (props.focusable)
        inject.onFocusableItemRemove()
    })

    return () => h(CollectionItemSlot, {
      id: id.value,
      focusable: focusable.value,
      active: active.value,
      scope: scopeOkuRovingFocusGroup.value,
    }, {
      default: () => {
        return h(Primitive.span, {
          'tabindex': isCurrentTabStop.value ? 0 : -1,
          'data-orientation': inject.orientation?.value,
          ...mergeProps(attrs, reactiveItemProps),
          'ref': forwardedRef,
          'onMousedown':
              composeEventHandlers<MouseEvent>((e) => {
                emit('mousedown', e)
              }, (event) => {
                // We prevent focusing non-focusable items on `mousedown`.
                // Even though the item has tabIndex={-1}, that only means take it out of the tab order.
                if (!focusable.value)
                  event.preventDefault()
                // Safari doesn't focus a button when clicked so we run our logic on mousedown also
                else inject.onItemFocus(id.value)
              }),
          'onFocus': composeEventHandlers<FocusEvent>((e) => {
            emit('focus', e)
          }, () => {
            inject.onItemFocus(id.value)
          }),
          'onKeydown': composeEventHandlers<KeyboardEvent>((e) => {
            emit('keydown', e)
          }, (event) => {
            if (event.key === 'Tab' && event.shiftKey) {
              inject.onItemShiftTab()
              return
            }

            if (event.target !== event.currentTarget)
              return

            const focusIntent = getFocusIntent(event, inject.orientation?.value, inject.dir?.value)

            if (focusIntent !== undefined) {
              event.preventDefault()

              const items = getItems().filter(item => item.focusable)
              let candidateNodes = items.map(item => item.ref.value)
              if (focusIntent === 'last') {
                candidateNodes.reverse()
              }
              else if (focusIntent === 'prev' || focusIntent === 'next') {
                if (focusIntent === 'prev')
                  candidateNodes.reverse()
                const currentIndex = candidateNodes.indexOf(event.currentTarget as HTMLElement)
                candidateNodes = inject.loop?.value
                  ? wrapArray(candidateNodes, currentIndex + 1)
                  : candidateNodes.slice(currentIndex + 1)
              }

              // /**
              //  * Imperative focus during keydown is risky so we prevent React's batching updates
              //  * to avoid potential bugs. See: https://github.com/facebook/react/issues/20332
              //  */
              setTimeout(() => focusFirst(candidateNodes))
            }
          }),
        }, slots.default?.())
      },
    })
  },

})

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuRovingFocusGroupItem = rovingFocusGroupItem as typeof rovingFocusGroupItem &
  (new () => {
    $props: RovingFocusGroupItemNaviteElement
  })
