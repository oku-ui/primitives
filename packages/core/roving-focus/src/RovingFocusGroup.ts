import { defineComponent, h, mergeProps } from 'vue'
import { useForwardRef, useListeners } from '@oku-ui/use-composable'
import { primitiveProps } from '@oku-ui/primitive'
import { OkuRovingFocusGroupImpl } from './RovingFocusGroupImpl'
import type { RovingFocusGroupNaviteElement } from './props'
import { CollectionItemSlot, CollectionProvider, CollectionSlot, GROUP_NAME, rovingFocusGroupProps, scopedProps } from './props'

const rovingFocusGroup = defineComponent({
  name: GROUP_NAME,
  components: {
    OkuRovingFocusGroupImpl,
    CollectionProvider,
    CollectionSlot,
    CollectionItemSlot,
  },
  inheritAttrs: false,
  props: {
    ...rovingFocusGroupProps.props,
    ...scopedProps,
    ...primitiveProps,
  },
  emits: rovingFocusGroupProps.emits,
  setup(props, { slots, attrs }) {
    const forwardedRef = useForwardRef()

    const emits = useListeners()
    return () => {
      return h(CollectionProvider, {
        scope: props.scopeOkuRovingFocusGroup,
      }, {
        default: () => h(CollectionSlot, {
          scope: props.scopeOkuRovingFocusGroup,
        }, {
          default: () => h(OkuRovingFocusGroupImpl, {
            ...mergeProps(attrs, props, emits),
            ref: forwardedRef,
          }, slots.default?.()),
        }),
      })
    }
  },
})

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuRovingFocusGroup = rovingFocusGroup as typeof rovingFocusGroup &
  (new () => {
    $props: RovingFocusGroupNaviteElement
  })
