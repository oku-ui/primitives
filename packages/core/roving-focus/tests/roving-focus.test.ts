import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'

import type { Component } from 'vue'
import { h } from 'vue'
import { OkuRovingFocusGroup, OkuRovingFocusGroupItem } from '../src'

const component = {
  setup(props, { slots, attrs }) {
    return () => h(OkuRovingFocusGroup, attrs, slots)
  },
} as Component

describe('okuRovingFocusGroup', () => {
  describe('okuRovingFocusGroupItem aschild', () => {
    it('empty', () => {
      const com = mount(component, {
        slots: {
          default: () => [
            h(OkuRovingFocusGroupItem, {
              asChild: true,
            }),
            h(OkuRovingFocusGroupItem, {
              asChild: true,
            }),
          ],
        },
      })
      expect(com.element).toMatchSnapshot()
    })

    it('one button', async () => {
      const com = mount(component, {
        slots: {
          default: () => [
            h(OkuRovingFocusGroupItem, {
              asChild: true,
            }, {
              default: () => h('button', {}, 'button'),
            }),
          ],
        },
      })
      expect(com.element).toMatchSnapshot()
    })
  })
})
