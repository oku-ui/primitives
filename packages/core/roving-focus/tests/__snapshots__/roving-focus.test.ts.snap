// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OkuRovingFocusGroup > OkuRovingFocusGroupItem aschild > empty 1`] = `
<div
  data-v-app=""
>
  
  <div
    style="outline-color: none; outline-style: none; outline-width: initial;"
    tabindex="-1"
  >
    <!---->
    <!---->
  </div>
  
</div>
`;

exports[`OkuRovingFocusGroup > OkuRovingFocusGroupItem aschild > one button 1`] = `
<div
  data-v-app=""
>
  
  <div
    style="outline-color: none; outline-style: none; outline-width: initial;"
    tabindex="-1"
  >
    <button
      data-oku-collection-item=""
      tabindex="-1"
    >
      button
    </button>
  </div>
  
</div>
`;
